// @chakra-ui/layout/dist/esm/divider.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import { chakra, forwardRef } from "@chakra-ui/system";
import { cx, __DEV__ } from "@chakra-ui/utils";
import * as React from "react";
/**
 * Layout component used to visually separate content in a list or group.
 * It display a thin horizontal or vertical line, and renders a `hr` tag.
 *
 * @see Docs https://chakra-ui.com/components/divider
 */

export var Divider = /*#__PURE__*/forwardRef(function Divider(props, ref) {
  var {
    orientation = "horizontal"
  } = props,
      rest = _objectWithoutPropertiesLoose(props, ["className", "orientation"]);

  var styles = {
    vertical: {
      borderLeftWidth: "1px",
      height: "100%"
    },
    horizontal: {
      borderBottomWidth: "1px",
      width: "100%"
    }
  };
  return /*#__PURE__*/React.createElement(chakra.hr, _extends({
    ref: ref,
    role: "separator",
    "aria-orientation": orientation
  }, rest, {
    __css: _extends({
      border: "0",
      opacity: 0.6,
      borderColor: "inherit",
      borderStyle: "solid"
    }, styles[orientation]),
    className: cx("chakra-divider", props.className)
  }));
});

if (__DEV__) {
  Divider.displayName = "Divider";
}
