// @chakra-ui/layout/dist/esm/list.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import { Icon } from "@chakra-ui/icon";
import { chakra, forwardRef } from "@chakra-ui/system";
import { getValidChildren, __DEV__ } from "@chakra-ui/utils";
import * as React from "react";

/**
 * List is used to display list items, it renders a `<ul>` by default.
 *
 * @see Docs https://chakra-ui.com/components/list
 */
export var List = /*#__PURE__*/forwardRef(function List(props, ref) {
  var {
    children,
    styleType = "none",
    stylePosition,
    spacing
  } = props,
      rest = _objectWithoutPropertiesLoose(props, ["children", "styleType", "stylePosition", "spacing"]);

  var validChildren = getValidChildren(children);
  var selector = "& > *:not(style) ~ *:not(style)";
  return /*#__PURE__*/React.createElement(chakra.ul, _extends({
    ref: ref,
    listStyleType: styleType,
    listStylePosition: stylePosition
    /**
     * We added this role to fix the Safari accessibility issue with list-style-type: none
     * @see https://www.scottohara.me/blog/2019/01/12/lists-and-safari.html
     */
    ,
    role: "list",
    __css: spacing ? {
      [selector]: {
        mt: spacing
      }
    } : {}
  }, rest), validChildren);
});

if (__DEV__) {
  List.displayName = "List";
}

export var OrderedList = /*#__PURE__*/forwardRef(function OrderedList(props, ref) {
  var rest = _objectWithoutPropertiesLoose(props, ["as"]);

  return /*#__PURE__*/React.createElement(List, _extends({
    ref: ref,
    as: "ol",
    styleType: "decimal",
    marginLeft: "1em"
  }, rest));
});

if (__DEV__) {
  OrderedList.displayName = "OrderedList";
}

export var UnorderedList = /*#__PURE__*/forwardRef(function UnorderedList(props, ref) {
  var rest = _objectWithoutPropertiesLoose(props, ["as"]);

  return /*#__PURE__*/React.createElement(List, _extends({
    ref: ref,
    as: "ul",
    styleType: "initial",
    marginLeft: "1em"
  }, rest));
});

if (__DEV__) {
  UnorderedList.displayName = "UnorderedList";
}

/**
 * ListItem
 *
 * Used to render a list item
 */
export var ListItem = chakra("li");

if (__DEV__) {
  ListItem.displayName = "ListItem";
}
/**
 * ListIcon
 *
 * Used to render an icon beside the list item text
 */


export var ListIcon = /*#__PURE__*/forwardRef(function ListIcon(props, ref) {
  return /*#__PURE__*/React.createElement(Icon, _extends({
    ref: ref,
    role: "presentation",
    marginRight: "0.5rem",
    display: "inline",
    verticalAlign: "text-bottom"
  }, props));
});

if (__DEV__) {
  ListIcon.displayName = "ListIcon";
}
