// @chakra-ui/transition/dist/cjs/collapse.js
"use strict";

exports.__esModule = true;
exports.Collapse = void 0;

var _utils = require("@chakra-ui/utils");

var _framerMotion = require("framer-motion");

var React = _interopRequireWildcard(require("react"));

var _utils2 = require("./__utils");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var variants = {
  exit: function exit(props) {
    return _extends({}, props.animateOpacity && {
      opacity: parseInt(props.startingHeight, 10) > 0 ? 1 : 0
    }, {
      height: props.startingHeight,
      transition: {
        duration: 0.2,
        ease: _utils2.EASINGS.easeInOut
      }
    });
  },
  enter: function enter(props) {
    return _extends({}, props.animateOpacity && {
      opacity: 1
    }, {
      height: props.endingHeight,
      transition: {
        duration: 0.3,
        ease: _utils2.EASINGS.easeInOut
      }
    });
  }
};
var Collapse = /*#__PURE__*/React.forwardRef(function Collapse(props, ref) {
  var isOpen = props["in"],
      unmountOnExit = props.unmountOnExit,
      _props$animateOpacity = props.animateOpacity,
      animateOpacity = _props$animateOpacity === void 0 ? true : _props$animateOpacity,
      _props$startingHeight = props.startingHeight,
      startingHeight = _props$startingHeight === void 0 ? 0 : _props$startingHeight,
      _props$endingHeight = props.endingHeight,
      endingHeight = _props$endingHeight === void 0 ? "auto" : _props$endingHeight,
      style = props.style,
      className = props.className,
      _onAnimationComplete = props.onAnimationComplete,
      rest = _objectWithoutPropertiesLoose(props, ["in", "unmountOnExit", "animateOpacity", "startingHeight", "endingHeight", "style", "className", "onAnimationComplete"]);

  var _React$useState = React.useState(function () {
    // If it's open by default, no need to apply `aria-hidden`
    if (isOpen) return false; // If startingHeight > 0, then content is partially visible

    if (parseInt(props.startingHeight, 10) > 0) return false; // Else, the content is hidden

    return true;
  }),
      ariaHidden = _React$useState[0],
      setAriaHidden = _React$useState[1];
  /**
   * Warn 🚨: `startingHeight` and `unmountOnExit` are mutually exclusive
   *
   * If you specify a starting height, the collapsed needs to be mounted
   * for the height to take effect.
   */


  if (startingHeight > 0 && unmountOnExit) {
    (0, _utils.warn)("startingHeight and unmountOnExit are mutually exclusive. You can't use them together");
  }

  var shouldExpand = unmountOnExit ? isOpen && unmountOnExit : true;
  var custom = {
    startingHeight: startingHeight,
    endingHeight: endingHeight,
    animateOpacity: animateOpacity
  };
  return /*#__PURE__*/React.createElement(_framerMotion.AnimatePresence, {
    initial: false,
    custom: custom
  }, shouldExpand && /*#__PURE__*/React.createElement(_framerMotion.motion.div, _extends({
    ref: ref,
    "aria-hidden": ariaHidden ? "true" : undefined,
    onAnimationComplete: function onAnimationComplete() {
      setAriaHidden(function (c) {
        return !c;
      });
      _onAnimationComplete == null ? void 0 : _onAnimationComplete();
    },
    className: (0, _utils.cx)("chakra-collapse", className),
    initial: "exit",
    animate: isOpen || unmountOnExit ? "enter" : "exit",
    exit: "exit"
  }, rest, {
    variants: variants,
    style: _extends({
      overflow: "hidden"
    }, style),
    custom: custom
  })));
});
exports.Collapse = Collapse;

if (_utils.__DEV__) {
  Collapse.displayName = "Collapse";
}
