// @chakra-ui/transition/dist/esm/collapse.js
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import { cx, warn, __DEV__ } from "@chakra-ui/utils";
import { AnimatePresence, motion } from "framer-motion";
import * as React from "react";
import { EASINGS } from "./__utils";
var variants = {
  exit: props => _extends({}, props.animateOpacity && {
    opacity: parseInt(props.startingHeight, 10) > 0 ? 1 : 0
  }, {
    height: props.startingHeight,
    transition: {
      duration: 0.2,
      ease: EASINGS.easeInOut
    }
  }),
  enter: props => _extends({}, props.animateOpacity && {
    opacity: 1
  }, {
    height: props.endingHeight,
    transition: {
      duration: 0.3,
      ease: EASINGS.easeInOut
    }
  })
};
export var Collapse = /*#__PURE__*/React.forwardRef(function Collapse(props, ref) {
  var {
    in: isOpen,
    unmountOnExit,
    animateOpacity = true,
    startingHeight = 0,
    endingHeight = "auto",
    style,
    className,
    onAnimationComplete: _onAnimationComplete
  } = props,
      rest = _objectWithoutPropertiesLoose(props, ["in", "unmountOnExit", "animateOpacity", "startingHeight", "endingHeight", "style", "className", "onAnimationComplete"]);

  var [ariaHidden, setAriaHidden] = React.useState(() => {
    // If it's open by default, no need to apply `aria-hidden`
    if (isOpen) return false; // If startingHeight > 0, then content is partially visible

    if (parseInt(props.startingHeight, 10) > 0) return false; // Else, the content is hidden

    return true;
  });
  /**
   * Warn 🚨: `startingHeight` and `unmountOnExit` are mutually exclusive
   *
   * If you specify a starting height, the collapsed needs to be mounted
   * for the height to take effect.
   */

  if (startingHeight > 0 && unmountOnExit) {
    warn("startingHeight and unmountOnExit are mutually exclusive. You can't use them together");
  }

  var shouldExpand = unmountOnExit ? isOpen && unmountOnExit : true;
  var custom = {
    startingHeight,
    endingHeight,
    animateOpacity
  };
  return /*#__PURE__*/React.createElement(AnimatePresence, {
    initial: false,
    custom: custom
  }, shouldExpand && /*#__PURE__*/React.createElement(motion.div, _extends({
    ref: ref,
    "aria-hidden": ariaHidden ? "true" : undefined,
    onAnimationComplete: () => {
      setAriaHidden(c => !c);
      _onAnimationComplete == null ? void 0 : _onAnimationComplete();
    },
    className: cx("chakra-collapse", className),
    initial: "exit",
    animate: isOpen || unmountOnExit ? "enter" : "exit",
    exit: "exit"
  }, rest, {
    variants: variants,
    style: _extends({
      overflow: "hidden"
    }, style),
    custom: custom
  })));
});

if (__DEV__) {
  Collapse.displayName = "Collapse";
}
