// @chakra-ui/system/dist/esm/providers.js
import { useColorMode } from "@chakra-ui/color-mode";
import { css } from "@chakra-ui/styled-system";
import { createContext, memoizedGet as get, mergeWith, runIfFn } from "@chakra-ui/utils";
import { Global, ThemeContext } from "@emotion/core";
import * as React from "react";
export var ThemeProvider = props => {
  var {
    children,
    theme
  } = props;
  var outerTheme = React.useContext(ThemeContext);
  var mergedTheme = mergeWith({}, outerTheme, theme);
  return /*#__PURE__*/React.createElement(ThemeContext.Provider, {
    value: mergedTheme
  }, children);
};
export function useTheme() {
  var theme = React.useContext(ThemeContext);

  if (!theme) {
    throw Error("useTheme: `theme` is undefined. Seems you forgot to wrap your app in `<ChakraProvider />`");
  }

  return theme;
}
var [StylesProvider, useStyles] = createContext({
  name: "StylesContext",
  errorMessage: "useStyles: `styles` is undefined. Seems you forgot to wrap the components in `<StylesProvider />` "
});
export { StylesProvider, useStyles };
/**
 * Applies styles defined in `theme.styles.global` globally
 * using emotion's `Global` component
 */

export var GlobalStyle = () => {
  var {
    colorMode
  } = useColorMode();
  return /*#__PURE__*/React.createElement(Global, {
    styles: theme => {
      var styleObjectOrFn = get(theme, "styles.global");
      var globalStyles = runIfFn(styleObjectOrFn, {
        theme,
        colorMode
      });
      if (!globalStyles) return undefined;
      var styles = css(globalStyles)(theme);
      return styles;
    }
  });
};
